### DISCUSSION TITLE
keyboardlayoutoverrides not working for me (trying to get colemak working)

### DISCUSSION THREAD

**Discussion by pupto-bismol** [tridactyl/tridactyl#4680] [Q&A]
here is what my current tridactylrc looks like:
```
set keyboardlayoutforce true
set keyboardlayoutoverrides.KeyN: hH
set keyboardlayoutoverrides.KeyE: jJ
set keyboardlayoutoverrides.KeyI: kK
set keyboardlayoutoverrides.KeyO: lL
set keyboardlayoutoverrides.KeyH: iI
set keyboardlayoutoverrides.KeyJ: nN
set keyboardlayoutoverrides.KeyK: eE
set keyboardlayoutoverrides.KeyL: oO
```
keyboardlayout force is indeed set to true when i tab complete it
when i try to tab complete keyboardlayoutoverrides, i get this put in my command bar:
`# keyboardlayoutoverrides {"KeyN:":"hH","KeyE:":"jJ","KeyI:":"kK","KeyO:":"lL","KeyH:":"iI","KeyJ:":"nN","KeyK:":"eE","KeyL:":"oO"}`

regardless of all of this, none of these binds work and nothing has changed. i even tried copy-pasting each `set keyboardlayoutoverrides...` command one by one into the command bar and it didn't change anything. exiting and restarting firefox doesn't change anything

maybe this has something to do with the bug listed in the help section? shown here:
"NB: due to a Tridactyl bug, you cannot set this using array notation as you can for, e.g. [[homepage]]. You must instead set the lower and upper registers using a string with no spaces in it, for example :set keyboardlayoutoverrides Digit2: 2" for the British English layout."

for some reason, even when i try to set the overrides command-by-command the tab completion leads me to believe it's interpreted as an array instead, which then causes it to work. but i'm very unsure as the documentation for changing the keyboard layout is incredibly confusing and lacks many useful examples.

**Comment by treapster** [comment]
I looked into it and it's a nasty bug caused by line https://github.com/tridactyl/tridactyl/blob/b4bb596de89b20bb067b26a3afa91eb4d18a7838/src/lib/keyseq.ts#L498 Thanks for reporting! It seems any updates stop propagating after the first override is added. What you could do is make sure keyboardlayoutoverrides is empty and put all override commands in .tridactylrc. From there it should be able to source it all at once.

**Reply by pupto-bismol** [reply]
when you say put all override commands in tridactylrc, that means put them as an array? and then sanitize everything and source?

**Reply by treapster** [reply]
If you just copy the separate commands as quoted in you post, it should work. But you can try array syntax too.

**Reply by pupto-bismol** [reply]
but putting them in my tridactylrc as shown and then sourcing was exactly how i set them the first time. they should have been set all at once - so i would think it would have worked then. i will try it again though
- just to note, i'm actually using ~/.config/tridactyl/tridactylrc as my config file instead of just ~/.tridactylrc, although i assume the functionality is the exact same

**Reply by pupto-bismol** [reply]
in order i ran these commands:

`sanitize tridactyllocal tridactylsync`

`source`

contents of ~/.config/tridactyl/tridactylrc:
```
unbind <C-f>
set keyboardlayoutforce true
set keyboardlayoutoverrides.KeyN: hH
set keyboardlayoutoverrides.KeyE: jJ
set keyboardlayoutoverrides.KeyI: kK
set keyboardlayoutoverrides.KeyO: lL
set keyboardlayoutoverrides.KeyH: iI
set keyboardlayoutoverrides.KeyJ: nN
set keyboardlayoutoverrides.KeyK: eE
set keyboardlayoutoverrides.KeyL: oO
```
unbinding <c-f> did work as expected, and the variables are set correctly again (tab completing keyboardlayoutoverrides gives the same array) but the overrides themselves didn't actually rebind any keys.

is there possibly an issue with the syntax of `keyboardlayoutoverrides.KeyX` here? the only reason i'm using it like this is because i saw it in a different random config file someone had hosted elsewhere, and i'm actually unsure of what the difference is. but i know that when i had the periods replaced with spaces it was giving errors as well as not working

**Reply by treapster** [reply]
There is other way - disable keyboardlayoutforce, then reopen tridactyl, then source, then enable:) We need to source config before local KEYCODETRANSLATEMAP is populated in the code, and fortunately we can do it by first changing config and then enabling keyboardlayoutforce. Sorry for the inconvenience, but i hope this workaround will work:)

**Reply by treapster** [reply]
> is there possibly an issue with the syntax of keyboardlayoutoverrides.KeyX here?

No, it's a bug with translation map caching
Or at least it looks very much like it

**Reply by pupto-bismol** [reply]
tried this:
`sanitize tridactyllocal tridactylsync`
(at this point verified keyboardlayoutforce is indeed set to false)
commented out keboardlayoutforce line on my tridactylrc, only things remaining in it are the <c-f> unbind and the overrides
`source`
`set keyboardlayoutforce true`

still does not work. again tab completion for keyboardlayoutforce and keyboardlayoutoverrides suggest they are set to the correct thing
(apologies if this isn't what you were telling me to try, this is just verifying what i did in case it was wrong)

**Reply by treapster** [reply]
Did you restart Firefox between disabling keyboardlayoutforce and sourcing?

**Reply by pupto-bismol** [reply]
yeah just tried that as well, no change. for reference i went to about:profiles and clicked "restart normally"
if this is a cache issue is there a cache file/directory i can delete?

**Reply by treapster** [reply]
There is a cache variable KEYCODETRANSLATEMAP(on the line linked on my first reply)which seems to be populated on the first keypress only, so it requires overrides to be set at that point. But as I said below it's weird because I am able to get the first override working(after I press a lot of keys), so something else is going on here

**Reply by pupto-bismol** [reply]
for me, even if i try to set just one override it doesn't change anything at all. that's after restarting firefox as well.
for example i did:
`sanitize tridactyllocal tridactylsync`
(restart firefox)
`set keyboardlayoutoverrides.keyE: jJ`
`set keyboardlayoutforce true`

as well as:
`sanitize tridactyllocal tridactylsync`
`set keyboardlayoutoverrides.keyE: jJ`
(restart firefox)
`set keyboardlayoutforce true`

and neither of these change the behaviour of e or E.

**Reply by treapster** [reply]
Not-working-at-all behavior at least makes sense given the code. The only thing left to try is reset config and then restart and source. But if you already tried it, I guess you'll have to wait a bit for a fix:(
Also, you said you're using ~/.config/tridactyl/tridactylrc. Are you passing it as an argument to `source`?

**Reply by pupto-bismol** [reply]
no, but it gets passed automatically anyways, other settings get changed just fine. i also have just tried it with a file at ~/.tridactylrc as well and the exact same thing happened

**Comment by treapster** [comment]
But the bug itself is even more mysterious, because with current code it should stop working as soon as the first key is pressed and KEYCODETRANSLATEMAP is populated, but for me it even makes the first override work somehow.

**Reply by bovine3dom** [reply]
We just need to add a config change listener here https://github.com/tridactyl/tridactyl/blob/b4bb596de89b20bb067b26a3afa91eb4d18a7838/src/lib/keyseq.ts#L521

**Reply by treapster** [reply]
Sure, but I tried to figure out why it doesn't work when working on the layout patches and couldn't determine the issue. And are you also having this weird only-first-override issue if you just use :set without reloading or anything?

**Reply by treapster** [reply]
Now that i've got some more time i can't even reproduce only-first-override bug. I got basically everything wrong with this issue, so nevermind:)

**Comment by bovine3dom** [comment]
There should not be any colons, is that written somewhere in the help or is it just a typo?

```
set keyboardlayoutforce true
set keyboardlayoutoverrides.KeyN hH
set keyboardlayoutoverrides.KeyE jJ
set keyboardlayoutoverrides.KeyI kK
set keyboardlayoutoverrides.KeyO lL
set keyboardlayoutoverrides.KeyH iI
set keyboardlayoutoverrides.KeyJ nN
set keyboardlayoutoverrides.KeyK eE
set keyboardlayoutoverrides.KeyL oO
```

**Reply by pupto-bismol** [reply]
oh wow thanks this worked
yeah i had just copied those definitions from a different rc file because the :help documentation doesn't really list any examples that work
